pkg/cmd/pulumi/package_add.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+	"github.com/spf13/cobra"
+)
+
+// Constructs the `pulumi package add` command.
+func newPackageAddCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "add <provider> [provider parameters]",
+		Args:  cobra.MinimumNArgs(1),
+		Short: "Add a package to your Pulumi project",
+		Long:  "Add a package to your Pulumi project.",
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+			proj, root, err := readProject()
+			if err != nil && errors.Is(err, workspace.ErrProjectNotFound) {
+				return errors.New("pulumi package add must be run in a Pulumi project")
+			}
+
+			ctx := cmd.Context()
+
+			plugin := args[0]
+			parameters := args[1:]
+
+			var pluginInstallCommand pluginInstallCmd
+			err = pluginInstallCommand.Run(ctx, []string{
+				"resource", /*plugin kind*/
+				plugin,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to install plugin: %w", err)
+			}
+
+			pkg, err := schemaFromSchemaSource(ctx, plugin, parameters)
+			if err != nil {
+				return fmt.Errorf("failed to get schema: %w", err)
+			}
+
+			language := proj.Runtime.Name()
+			out := root
+			local := true
+
+			err = genSDK(
+				language,
+				out,
+				pkg,
+				"",    /*overlays*/
+				local, /*local*/
+			)
+			if err != nil {
+				return fmt.Errorf("failed to generate SDK: %w", err)
+			}
+
+			printLinkInstructions(language, plugin, out)
+
+			return nil
+		}),
+	}
+
+	return cmd
+}
+
+// Prints instructions for linking a locally generated SDK to an existing
+// project, in the absence of us attempting to perform this linking automatically.
+func printLinkInstructions(language string, plugin string, out string) {
+	switch language {
+	case "nodejs":
+		printNodejsLinkInstructions(plugin, out)
+	case "python":
+		printPythonLinkInstructions(plugin, out)
+	case "go":
+		printGoLinkInstructions(plugin, out)
+	case "dotnet":
+		printDotnetLinkInstructions(plugin, out)
+	case "java":
+		printJavaLinkInstructions(plugin, out)
+	default:
```

Comment:
I have the bad habit of not thinking about `yaml`. Are param providers for yaml still undecided or do we know what we will do for them?

Reply:
Ach excellent point! Come to think of it, what does `gen-sdk` even do for YAML? Does it work but no-op? Will test.

Reply:
Oh good question! I guess in the future we might write that lock file with the param somewhere, but for now there's probably nothing? or a panic? :D

Reply:
Note: `gen-sdk` with YAML currently errors, so I've added a short-circuit in `package add` to try and avoid us hitting this.